jsonnet/prometheus.jsonnet

```diff
@@ -382,7 +382,7 @@ local metrics = import 'telemeter-client/metrics.jsonnet';
                       values: [$._config.prometheus.name],
                     }]
                   },
-                  topologyKey: 'kubernetes.io/hostname',
+                  topologyKey: 'topology.kubernetes.io/zone',
```

Comment:
It is a valid point but we need to ensure that the notion of availability is clearly understood. It can be at zone level or node level. What does CMO prefer? At a high level, the knowledge of where the replicas should be and the number of replicas should be within CMO code so that the decisions can be made dynamically. We can add both the labels for now but I think the decision should be made by CMO team.

> Also, I still think we should keep the hard-affinity on hostname to prevent scheduling both Prometheus instance on the same node as it causes a couple of issues such as https://bugzilla.redhat.com/show_bug.cgi?id=1955147

If all the nodes are in the same zone, we will have only pod running and CMO would fail to reconcile the 2nd pod and the upgrade would fail, we would not see the issue mentioned in the BZ.

Reply:
How I see it is that CMO should have hard-affinity on hostname and soft-affinity on zones since there are no guarantees that the cluster is across multiple zones and we always want to make sure that replicas are not scheduled on the same node to avoid having a single point of failure.

That said, would you expect some logic in CMO to adjust the affinity constraints on the zones dynamically based on how many zones are available for the cluster?

> If all the nodes are in the same zone, we will have only pod running and CMO would fail to reconcile the 2nd pod and the upgrade would fail, we would not see the issue mentioned in the BZ.

Yes, that's the problem here. If we were to enforce hard-affinity on zone as is, the Prometheus instance would become unschedulable if all nodes are all on the same zone. Then comes the question of how should we handle the case of single-zone clusters? Is that even a thing? Are we in OpenShift requiring nodes to be in at least 2 zones in order to enable HA? Kind of like the constraint we have for nodes, but on zones.

Reply:
> Is that even a thing?

It's a thing.  The CI AWS account doesn't have access to any regions with a single available zone at the moment:

```console
$ aws --region us-east-1 ec2 describe-regions --output json | jq -r '.Regions[].RegionName' | while read REGION; do ZONES="$(aws --region "${REGION}" ec2 describe-availability-zones --output json | jq -r '[.AvailabilityZones[] | select(.State == "available")] | length')"; echo "${ZONES} ${REGION}"; done | sort -n
2 us-west-1
3 ap-northeast-1
...
```

But zones can have outages, and the installer will not try to set up machines in zones that are currently down.  I can see what Telemetry has to say about how often clusters actually do this...

Reply:
I don't think we have this information in Telemetry yet. We have `cluster_infrastructure_provider` that gives information about the cloud provider and the region, but nothing about the availability zones.

The only place that I am aware of where we could get this information is from the `label_topology_kubernetes_io_zone` label of the `kube_node_label` metrics, but after checking, we are not populating it anywhere in the monitoring stack.